### PR TITLE
KAFKA-5445-1: document exceptions for create, list, delete, describe

### PR DESCRIPTION
**Pull request by adyach**
Exceptions are processed internally in KafkaAdminClient without throwing them to the client code, hence the documentation of the exception is done in unusual way.

The PR is the follow up for #4000 which is gonna be split in 4 different PRs.

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @cmccabe

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java [hunk 1]
```diff
@@ -95,6 +104,29 @@ public void close() {
...
+     * {@link CreateTopicsResult#values() values()} method of the  returned {@code CreateTopicsResult}</p>
+     * <ul>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidRequestException}
+     *     if duplicate topics were present in the request</li>
+     *     <li>{@link org.apache.kafka.common.errors.ClusterAuthorizationException}
+     *     if cluster authorization failed for create operation</li>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidPartitionsException}
+     *     if number of partitions is less than 1</li>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidReplicationFactorException}
+     *     if replication factor is less than 1 or should be large than number of available brokers</li>
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java]: Typo: should be "or larger than the number of available brokers"

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java [hunk 2]
```diff
@@ -95,6 +104,29 @@ public void close() {
      *
      * This operation is supported by brokers with version 0.10.1.0 or higher.
      *
+     * <p>The following exceptions can be anticipated when calling {@code get()} on the futures obtained from the
+     * {@link CreateTopicsResult#values() values()} method of the  returned {@code CreateTopicsResult}</p>
+     * <ul>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidRequestException}
+     *     if duplicate topics were present in the request</li>
+     *     <li>{@link org.apache.kafka.common.errors.ClusterAuthorizationException}
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java]: I do not think this exception can be thrown from CreateTopics.

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java [hunk 3]
```diff
@@ -95,6 +104,29 @@ public void close() {
...
+     *     <li>{@link org.apache.kafka.common.errors.InvalidPartitionsException}
+     *     if number of partitions is less than 1</li>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidReplicationFactorException}
+     *     if replication factor is less than 1 or should be large than number of available brokers</li>
+     *     <li>{@link org.apache.kafka.common.errors.TopicExistsException}
+     *     if topic name already exists</li>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidTopicException}
+     *     if topic name collides with other topic name</li>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidReplicaAssignmentException}
+     *     if the proposed replica assignment is invalid. For example if some of the partitions have different number
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java]: Should have a comma after "for example"

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java [hunk 4]
```diff
@@ -128,6 +160,15 @@ public abstract CreateTopicsResult createTopics(Collection<NewTopic> newTopics,
      *
      * This operation is supported by brokers with version 0.10.1.0 or higher.
      *
+     * <p>The following exceptions can be anticipated when calling {@code get()} on the futures obtained from the
+     * {@link DeleteTopicsResult#values() values()} method of the  returned {@code DeleteTopicsResult}</p>
+     * <ul>
+     *     <li>{@link org.apache.kafka.common.errors.UnknownTopicOrPartitionException}
+     *     if the topic does not exist</li>
+     *     <li>{@link org.apache.kafka.common.errors.NotControllerException}
+     *     if this is not the correct controller for the cluster</li>
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java]: This exception can't be thrown by DeleteTopics.

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java [hunk 5]
```diff
@@ -40,6 +40,15 @@
...
  * manner, but we reserve the right to make breaking changes in minor releases, if necessary. We will update the
  * {@code InterfaceStability} annotation and this notice once the API is considered stable.
+ *
+ * <p>The following exceptions can be thrown when calling {@code get()} on the futures obtained from AdminClient methods</p>
+ * <ul>
+ *     <li>{@link org.apache.kafka.common.errors.TimeoutException}
+ *     if the request was not completed in within the given {@link CreateTopicsOptions#timeoutMs()}</li>
+ *     <li>{@link org.apache.kafka.common.errors.TopicAuthorizationException}
+ *     if the authenticated user is not authorized to alter the topic</li>
+ * </ul>
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:52]: You should also include UnknownServerException here, which indicates an internal problem on the server side.

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java [hunk 6]
```diff
@@ -128,6 +160,15 @@ public abstract CreateTopicsResult createTopics(Collection<NewTopic> newTopics,
...
      * This operation is supported by brokers with version 0.10.1.0 or higher.
      *
+     * <p>The following exceptions can be anticipated when calling {@code get()} on the futures obtained from the
+     * {@link DeleteTopicsResult#values() values()} method of the  returned {@code DeleteTopicsResult}</p>
+     * <ul>
+     *     <li>{@link org.apache.kafka.common.errors.UnknownTopicOrPartitionException}
+     *     if the topic does not exist</li>
+     *     <li>{@link org.apache.kafka.common.errors.NotControllerException}
+     *     if this is not the correct controller for the cluster</li>
+     * </ul>
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:168]: Please include TopicDeletionDisabledException here.

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java [hunk 7]
```diff
@@ -163,6 +204,11 @@ public DeleteTopicsResult deleteTopics(Collection<String> topics) {
      * This is a convenience method for #{@link AdminClient#listTopics(ListTopicsOptions)} with default options.
      * See the overload for more details.
      *
+     * <ul>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidReplicationFactorException}
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java]: This exception can't be thrown by listTopics.

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java [hunk 8]
```diff
@@ -183,6 +229,15 @@ public ListTopicsResult listTopics() {
      * This is a convenience method for #{@link AdminClient#describeTopics(Collection, DescribeTopicsOptions)} with
      * default options. See the overload for more details.
      *
+     * <p>The following exceptions can be anticipated when calling {@code get()} on the futures obtained from the
+     * {@link DescribeTopicsResult#values() values()} method of the  returned {@code DescribeTopicsResult}</p>
+     * <ul>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidTopicException}
+     *     if topic name collides with other topic name</li>
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java]: InvalidTopicException happens when the topic name can't be represented in the request, or if it is not found, not if it collides with another topic name.

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java [hunk 9]
```diff
@@ -183,6 +229,15 @@ public ListTopicsResult listTopics() {
...
+     *     <li>{@link org.apache.kafka.common.errors.InvalidReplicationFactorException}
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java]: This exception can't happen in DescribeTopics.

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java [hunk 10]
```diff
@@ -95,6 +104,29 @@ public void close() {
...
+     *     if duplicate topics were present in the request</li>
+     *     <li>{@link org.apache.kafka.common.errors.ClusterAuthorizationException}
+     *     if cluster authorization failed for create operation</li>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidPartitionsException}
+     *     if number of partitions is less than 1</li>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidReplicationFactorException}
+     *     if replication factor is less than 1 or should be large than number of available brokers</li>
+     *     <li>{@link org.apache.kafka.common.errors.TopicExistsException}
+     *     if topic name already exists</li>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidTopicException}
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:118]: This exception happens if the given topic name can't be represented, not if it collides with another topic name.

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java [hunk 11]
```diff
@@ -95,6 +104,29 @@ public void close() {
      *
      * This operation is supported by brokers with version 0.10.1.0 or higher.
      *
+     * <p>The following exceptions can be anticipated when calling {@code get()} on the futures obtained from the
+     * {@link CreateTopicsResult#values() values()} method of the  returned {@code CreateTopicsResult}</p>
+     * <ul>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidRequestException}
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java]: This exception isn't used in createTopics.

**Comment by ijuma** [comment]
@adyach, can you please address the review comments?

**Comment by adyach** [comment]
@ijuma thank you for reminding! done.

**Comment by mjsax** [comment]
@cmccabe @ijuma What is the status of this PR?

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java [hunk 12]
```diff
@@ -183,6 +222,13 @@ public ListTopicsResult listTopics() {
      * This is a convenience method for #{@link AdminClient#describeTopics(Collection, DescribeTopicsOptions)} with
      * default options. See the overload for more details.
      *
+     * <p>The following exceptions can be anticipated when calling {@code get()} on the futures obtained from the
+     * {@link DescribeTopicsResult#values() values()} method of the  returned {@code DescribeTopicsResult}</p>
+     * <ul>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidTopicException}
+     *     if the topic name can't be represented in the request, or if it is not found</li>
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:229]: This is not correct.  We return `UnknownTopicOrPartitionException` if the topic is not found.

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java [hunk 13]
```diff
@@ -95,6 +106,25 @@ public void close() {
...
+     * {@link CreateTopicsResult#values() values()} method of the  returned {@code CreateTopicsResult}</p>
+     * <ul>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidPartitionsException}
+     *     if number of partitions is less than 1</li>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidReplicationFactorException}
+     *     if replication factor is less than 1 or larger than the number of available brokers</li>
+     *     <li>{@link org.apache.kafka.common.errors.TopicExistsException}
+     *     if topic name already exists</li>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidTopicException}
+     *     if the topic name can't be represented in the request, or if it is not found</li>
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:119]: We do not throw `InvalidTopicException` "if [the topic] is not found"

**Comment by cmccabe** [comment]
I left some comments.  Looks like it's getting closer.

**Pull request closed without merging**